### ISSUE TITLE
Assess the consequences and/or tradeoffs of multiple File entities referencing the same bytestream

### ISSUE THREAD

**Issue by emetsger** [jhu-idc/iDC-general#310]
The [implementation of media migrations](https://github.com/jhu-idc/idc-isle-dc/pull/95) introduced content addressed filenames, which simplifies reasoning about what Drupal does with filename collisions and greatly reduces the chances of accidentally overwriting an existing file with different content (additional benefits are that checking the integrity of the files are simplified because the asserted checksum is the name of the file itself).

> Note that content addressed filenames are _only_ implemented for `Media` that are ingested into Drupal by migrations.  Media ingested interactively will not use content addressing.  This issue assumes that migrations are being used to ingest Media and their files.

Separate, but related to content addressed filenames, is the `File` entity and its relationship to the bytestream it describes.  A relationship between a `File` and the bytestream is established whenever a `Media` is created.  At issue is the _cardinality_ of the relationship: may multiple `File` entities refer to the same bytestream, or ought they be limited to a one-to-one relationship.  The Drupal model allows for either, and the choice is ours to make.

For example, consider a [typical Media migration definition](https://github.com/jhu-idc/idc-isle-dc/blob/development/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml).  It [indicates](https://github.com/jhu-idc/idc-isle-dc/blob/development/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml#L61) that when a file is imported (thereby creating a `File` entity), we do _not_ reuse an existing `File` entity; instead, we create a new one.  (The introduction to content addressed files is relevant here, because Drupal determines whether or not a `File` entity exists based on the file name).

The goal of this issue is to assess the consequences of allowing multiple `File` entities to reference a single bytestream, or restricting the relationship to one `File` per bytestream.

(Note this issue was introduced with the implementation of [media migrations](https://github.com/jhu-idc/idc-isle-dc/pull/95))

**Comment by adam-vessey** [comment]
I could foresee some issues:
* [URI's on File entities are intended to be unique](https://git.drupalcode.org/project/drupal/-/blob/fd638d02d987528014b83cdd002ad91abfe10bfc/core/modules/file/src/Entity/File.php#L240-245):
    * You _can_ create them if you do so without validation; however, if you validate when saving the second (and later) copies, it will throw errors.
      <details>
      <summary>Example/test code</summary>
 
       ```php
      $uri = 'temporary://some-file.txt';
      file_put_contents($uri, 'Some contents.');

      $etm = \Drupal::entityTypeManager();
      $storage = $etm->getStorage('file');
  
      $files = [];
  
      try {
        foreach (['alpha', 'bravo', 'charlie'] as $name) {
          $file = $storage->create([
            'filename' => $name,
            'uri' => $uri,
          ]);
          $file->setValidationRequired(TRUE);
          $errors = $file->validate();
          if ($errors->count() === 0) {
            $file->save();
            $files[] = $file;
          }
          else {
            dsm("$errors");
          }
        }
      }
      finally {
        foreach ($files as $file) {
          $file->delete();
        }
      }
      ```
      </details>
* Because they're expected to be unique, [Drupal expects to delete the underlying file-on-disk when File entities are deleted](https://git.drupalcode.org/project/drupal/-/blob/fd638d02d987528014b83cdd002ad91abfe10bfc/core/modules/file/src/Entity/File.php#L205-213)
  <details>
  <summary>Example/test code</summary>

  ```php
  $uri = 'temporary://some-file.txt';
  file_put_contents($uri, 'Some contents.');
  
  $etm = \Drupal::entityTypeManager();
  $storage = $etm->getStorage('file');
  
  $files = [];
  
  try {
    foreach (['alpha', 'bravo', 'charlie'] as $name) {
      $file = $storage->create([
        'filename' => $name,
        'uri' => $uri,
      ]);
      $file->save();
      $files[] = $file;
    }
    
    foreach ($files as $file) {
      if (!file_exists($file->getFileUri())) {
        dsm("{$file->label()} points at something that no longer exists.");
      }
      else {
        dsm("{$file->label()} points at an existing file.");
      }
    }
  
    $to_delete = array_pop($files);
    $to_delete->delete();
    dsm("deleted {$to_delete->label()}");
  
    foreach ($files as $file) {
      if (!file_exists($file->getFileUri())) {
        dsm("{$file->label()} points at something that no longer exists.");
      }
      else {
        dsm("{$file->label()} points at an existing file.");
      }
    }
  
  }
  finally {
    foreach ($files as $file) {
      $file->delete();
    }
  }
  ```
  </details>

It seems like File Entities are intended to directly represent the files-on-disk/byte streams, and to allow for their management inside of Drupal. Something of an aside, but Drupal _does_ have a reference-counting mechanism (you can [see it clear out any entries when explicitly deleting the file](https://git.drupalcode.org/project/drupal/-/blob/fd638d02d987528014b83cdd002ad91abfe10bfc/core/modules/file/src/Entity/File.php#L198-204)). These reference counts might then be used for garbage collection; however, [this feature has been disabled by default](https://www.drupal.org/node/2891902).

The better approach may be to lookup and reference existing file entities based on hash (or whatever uniqueness criteria), such that multiple media entities could refer to the same underlying file entity, and therefore the same file-on-disk/bytestream, with any instance-specific metadata put to the media. [Double-checking the `file_import` code for the `reuse` parameter](https://git.drupalcode.org/project/migrate_file/-/blob/2.0.x/src/Plugin/migrate/process/FileImport.php#L221-242), it appears that it would facilitate this in the context of content-addressed file URIs.

**Comment by birkland** [comment]
Thanks!   @emetsger can you review this and recommend a course of action?  

It seems clear that a course correction will be needed.  The question, in my mind, is whether we simply revert to the default file naming provided by Drupal (i.e, not the hash) and 1:1 file-to-bytestream references, or if we want to spend effort trying to reference existing file entities, or <insert something else here that is not obvious to me right now>

@adam-vessey As far as the ` lookup and reference existing file entities based on hash` approach, what sort of consequences would that have on the requirements for writing CSV files for migrations?  (In particular, would our librarians and metadata experts need craft their migration CSVs in a way that works with this approach?)  For reference, an existing CSV for ingesting an image files to attach to a repository item here:  https://github.com/jhu-idc/idc-isle-dc/blob/development/tests/10-migration-backend-tests/testcafe/migrations/media-image.csv

**Comment by adam-vessey** [comment]
@birkland: No, CSVs should be able to remain as they are, it's more the migration implementation itself which might have to be adjusted, such as:

<details>
  <summary>Example diff</summary>
  
  ```diff
  diff --git a/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml b/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml
  index adcfc66..5609220 100644
  --- a/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml
  +++ b/codebase/config/sync/migrate_plus.migration.idc_ingest_media_file.yml
  @@ -57,16 +57,38 @@ process:
         - constants/DRUPAL_FS
         - '@_destination_filepath'
     field_original_name: original_name
  +  _file_lookup:
  +    # XXX: It appears that "filehash" does not actually make the hashes
  +    # queryable of the file entity... more just here to illustrate what things
  +    # might look like.
  +    - plugin: entity_lookup
  +      source: '@_file_sha'
  +      entity_type: file
  +      value_key: filehash
  +  _file_import:
  +    # Import if we failed to lookup.
  +    - plugin: skip_on_value
  +      source: '@_file_lookup'
  +      not_equal: true
  +      # XXX: migrate_plus's performs comparisons as strings... could always roll
  +      # another plugin to compare differently.
  +      value: null
  +    - plugin: file_import
  +      source: '@_download_filepath'
  +      move: true
  +      reuse: false
  +      rename: false
  +      id_only: true
  +      destination: '@_destination_drupalpath'
  +      mimetype: mime_type
  +      filename: name
     field_media_file/target_id:
  -    plugin: file_import
  -    source: '@_download_filepath'
  -    move: true
  -    reuse: false
  -    rename: false
  -    id_only: true
  -    destination: '@_destination_drupalpath'
  -    mimetype: mime_type
  -    filename: name
  +    # Use either the looked-up or imported file.
  +    - plugin: get
  +      source:
  +        - '@_file_lookup'
  +        - '@_file_import'
  +    - plugin: null_coalesce
     field_mime_type: mime_type
     name: name
     field_media_of:
   ```
   
 </details>

... so yeah, might require creating/extending a migrate process plugin or two (or the bigger task of making filehash make its properties queryable).

**Comment by birkland** [comment]
Great, thanks.  This is helpful!  Elliot gets back in the office Tuesday, so hopefully we can make a decision next week as far as what to do and whether the cost/benefit ratio is there for pursuing process plugins and queryable hashes.  I don't believe deduplication is one of our essential-for-launch use cases.  Using filehash seemed like potentially a cheap win at the time, but with possible consequences we weren't able to reason about.

**Comment by emetsger** [comment]
We avoid creating multiple `File` entities by switching the `reuse` parameter to `true` in the existing migration.  Beyond that, I don't think anything else needs to change.  If there's any pre-existing `File` with the same URI as the file being migrated, it will use it.

**Comment by adam-vessey** [comment]
Yeah, seems like it should do the trick, with the understanding that the deduping provided is being scoped/confined to just the domain of migrations.

**Comment by birkland** [comment]
Fix shall be implemented in #377
